### PR TITLE
feat: added error codes

### LINKED ISSUE #56: Request for v4 parity
* [x] Vertex mode - thanks to @mojixcoder 
* [ ] Error codes

### PR DESCRIPTION
**Pull request by mojixcoder**
Breaking changes are introduced to the package's API.
Closes #56

### CODE REVIEW

File: .github/workflows/build.yml
Comment (dfellis) [inline .github/workflows/build.yml:14]: Is there a reason you aren't keeping 1.20 in this list and dropping it entirely?

I don't develop in Go, so not sure what the community stance is on backwards compatibility in libraries.

File: go.mod
Comment (dfellis) [inline go.mod:3]: Again, why change the minimum version here? Are you using a feature that only exists in 1.21?

Comment (mojixcoder) [reply go.mod:3]: Not really, usually most of the libraries support 3 latest Go versions. Makes it easier for maintainers to maintain the package.
So it depends on h3-go’s backward comptability policies.
I can downgrade to Go 1.20 as well.

Comment (dfellis) [reply go.mod:3]: Got it. As I said, I don't develop in Go, so I will leave that decision to @zachcoleman and/or @jogly

File: h3.go
Comment (zachcoleman) [inline h3.go:89]: Think we should opt for `C.uint32_t` just to match what's explicitly in the C code... I think on the Go side they're the same, but seems like it'd be nice to match explicitly. https://github.com/uber/h3/blob/e8b16611b582275bdb9815bc5a80e0559a811758/src/h3lib/include/h3api.h.in#L79

File: h3.go
Comment (zachcoleman) [inline h3.go:150]: This happens a lot in the code right, but there's one interesting behavior w/ a map here. I think could be a risk here when not checking for the key in the map since the zero value of an error is `nil` and so if an error code is introduced, this would return `nil`.

**Comment by zachcoleman** [comment]
Overall, introducing breaking changes (even rightfully so), I personally want to think on for a bit. The repo versions with major and minor for upstream h3 ([ref](https://github.com/uber/h3-go/blob/d58e7c61ec519e697051f88b032e93f3d69fef64/CHANGELOG.md?plain=1#L6-L8)). So maybe this is fine, but definitely feels like this cutover will break a lot if users seemingly upgrade the patch version due to these semantics. @jogly Do you have an opinion on this?

**Comment by jogly** [comment]
I've hemmed and hawed on this for a while myself. I agree w your hesitation, and am unwilling to release breaking changes as a patch version for the reasons you mention. My latest thought is to publish this in a v4.2 prerelease branch/tag and wait until core is at 4.2 to make it a full release.

Another thought would be to consider an alternate strategy where we keep the package functions' signature as-is, and introduce an h3 struct with the signatures from this PR. Then the package functions could call an unexported default structs' functions and swallow the error for backwards compatibility and ergonomics.

File: go.mod
Comment (jogly) [reply go.mod:3]: I only care we support the versions of Go that Go cares to support (https://go.dev/doc/devel/release), and their policy is 2 versions. Go is at 1.23 now so I support dropping 1.20 😁

File: h3.go
Comment (mojixcoder) [reply h3.go:150]: I also thought about this, do you think H3 core would return other error codes?

File: h3.go
Comment (mojixcoder) [reply h3.go]: Done

File: h3.go
Comment (mojixcoder) [reply h3.go]: Done

**Comment by mojixcoder** [comment]
> I've hemmed and hawed on this for a while myself. I agree w your hesitation, and am unwilling to release breaking changes as a patch version for the reasons you mention. My latest thought is to publish this in a v4.2 prerelease branch/tag and wait until core is at 4.2 to make it a full release.

Why not a major release like v5?
This change introduces breaking changes to almost all functions. Do you want to keep h3-go versions same as h3 versions?

File: h3.go
Comment (zachcoleman) [reply h3.go]: @mojixcoder Thank you!

To your question, they've (h3 core) certainly given themselves space to add error codes w/ the uint here and I'm sure this set grew over time as well. Just a little future-proofing on our end.

**Comment by zachcoleman** [comment]
> > I've hemmed and hawed on this for a while myself. I agree w your hesitation, and am unwilling to release breaking changes as a patch version for the reasons you mention. My latest thought is to publish this in a v4.2 prerelease branch/tag and wait until core is at 4.2 to make it a full release.
> 
> Why not a major release like v5? This change introduces breaking changes to almost all APIs. Do you want to keep h3-go versions same as h3 versions?

That has been the standard to date in the repo: https://github.com/uber/h3-go/blob/d58e7c61ec519e697051f88b032e93f3d69fef64/CHANGELOG.md?plain=1#L6-L8

**Comment by zachcoleman** [comment]
I'm inclined to want to be patient and make this cutover on an h3-core 4.2 release. I reached out to h3-core team on Slack and they're inclined to release soon. I think a little patience here can go a long way and save some complicated work.

Ref: https://h3-core.slack.com/archives/CPU7D1TRD/p1728850733866769

**Comment by zachcoleman** [comment]
Thank you for your patience @mojixcoder... there's still no movement on the core library right now. I just wanted to poke this and let you know I'm checking frequently and we will act quickly after a core library releases and your contribution is very much appreciated ❤️.

**Comment by mojixcoder** [comment]
Thank you @zachcoleman ❤️. 
That’s fine you can merge it whenever you want

**Comment by mojixcoder** [comment]
Hello,
Just wanted to let you know that h3 core v4.2 got released 2 weeks ago.
Should we upgrade h3 deps before merging this? Or it’s not necessary.

**Comment by zachcoleman** [comment]
> Hello, Just wanted to let you know that h3 core v4.2 got released 2 weeks ago. Should we upgrade h3 deps before merging this? Or it’s not necessary.

Thanks for reaching out. I have been away for holidays. I will look into this soon and I think you are right. We should probably be running the shell script to update to the new upstream version and then resolving conflicts as necessary.

If you'd like for your PR that's a good first step to also incorporate the upstream changes and verify the test suite.

**Comment by zachcoleman** [comment]
@mojixcoder I have merged the version upgrade PR. Can you rebase this one and/or resolve the conflicts? 

I'll wait for a release for this commit 🥳

**Comment by mojixcoder** [comment]
Can you please do something so that CI runs without requiring approval, at least for me?
I'm not able to see the result of my changes immediately.

**Review by zachcoleman — APPROVED** [review]
looks good and looks like everything is passing 🥳

**Pull request merged**
